### PR TITLE
[not working] beginning of assigned IR functions (interval operations)

### PR DESCRIPTION
**Pull request by tpoterba**
I'm not going to be able finish my assignments before going on vacation 😱! I couldn't figure out how to make the `Code` generics work here, but thought I'd push it up because you'll probably solve it instantly.

### CODE REVIEW

**Comment by catoverdrive** [comment]
@jigold I added some things to this PR to actually do the AST conversion for some of these, so I'd appreciate another pair of eyes for review. This is mainly to handle the fact that some things (e.g. `Interval.start`, `Locus.contig`) that are currently handled as functions in IR are going through `Select` in AST.

File: src/main/scala/is/hail/expr/ir/functions/IntervalFunctions.scala [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+import is.hail.expr.{Apply, ir, types}
+import is.hail.expr.types.{TBoolean, TBooleanOptional, TInterval, TStruct}
+import is.hail.utils._
+
+object IntervalFunctions extends RegistryFunctions {
+
+  def registerAll(): Unit = {
+
+    registerCode("Interval", Array(tv("T"), tv("T"), TBoolean(), TBoolean()), TInterval(tv("T")), isDet = true) {
+      (mb, args) =>
```
Comment (jigold) [inline src/main/scala/is/hail/expr/ir/functions/IntervalFunctions.scala]: Can you add a test suite for these like I did for CallFunctions?

Comment (catoverdrive) [reply src/main/scala/is/hail/expr/ir/functions/IntervalFunctions.scala]: o7

File: src/main/scala/is/hail/expr/ir/functions/IntervalFunctions.scala [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+          srvb.start(),
+          srvb.addIRIntermediate(tv("T").t)(args(0)),
+          srvb.advance(),
+          srvb.addIRIntermediate(tv("T").t)(args(1)),
+          srvb.advance(),
+          srvb.addBoolean(args(2).asInstanceOf[Code[Boolean]]),
+          srvb.advance(),
+          srvb.addBoolean(args(3).asInstanceOf[Code[Boolean]]),
+          srvb.advance(),
+          srvb.offset
```
Comment (jigold) [inline src/main/scala/is/hail/expr/ir/functions/IntervalFunctions.scala]: Is there a preference for `srvb.offset` vs `srvb.end()`

Comment (catoverdrive) [reply src/main/scala/is/hail/expr/ir/functions/IntervalFunctions.scala]: I've been using `offset` everywhere. I've no idea if other people have preferences.

File: src/test/scala/is/hail/expr/ir/IntervalSuite.scala [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  val i2 = interval(point(1), NA(tpoint1), true, false)
+  val i3 = interval(NA(tpoint1), point(2), true, false)
+  val i4 = interval(NA(tpoint1), point(2), null, false)
+  val i5 = interval(NA(tpoint1), point(2), true, null)
+
+  @Test def constructor() {
+    assertEvalsTo(i1, Interval(Row(1), Row(2), true, false))
+    assertEvalsTo(i2, Interval(Row(1), null, true, false))
+    assertEvalsTo(i3, Interval(null, Row(2), true, false))
+    assertEvalsTo(i4, null)
```
Comment (jigold) [inline src/test/scala/is/hail/expr/ir/IntervalSuite.scala]: So start and end are allowed to be null but not the includes start and end?

Comment (catoverdrive) [reply src/test/scala/is/hail/expr/ir/IntervalSuite.scala]: yeah. TInterval has required booleans for includesStart and includesEnd.

**Pull request merged**